README.md
# ASUS Aura Sync for VSCode (vscode-aura)

[![Visual Studio Marketplace Version](https://img.shields.io/visual-studio-marketplace/v/obfuscatedgenerated.vscode-aura) ![Visual Studio Marketplace Installs](https://img.shields.io/visual-studio-marketplace/d/obfuscatedgenerated.vscode-aura?label=installs) ![Visual Studio Marketplace Rating (Stars)](https://img.shields.io/visual-studio-marketplace/stars/obfuscatedgenerated.vscode-aura)](https://marketplace.visualstudio.com/items?itemName=obfuscatedgenerated.vscode-aura)


Adds ASUS Aura RGB integration to VSCode. (Unofficial, Windows x64 Only)

ASUS Aura Sync® is either a US registered trademark or trademark of ASUSTeK Computer Inc. in the United States and/or other countries.

Reference to any ASUS products, services, processes, or other information and/or use of ASUS Trademarks does not constitute or imply endorsement, sponsorship, or recommendation thereof by ASUS.

> **Warning**:
Sometimes, new VSCode versions update Electron. When this happens, I need to manually recompile the native module. If you get an error about the module version not being correct, please open an issue and I'll get it fixed ASAP. This also drops support for older VSCode versions. If you know of a way I can prebuildify to support all Electron versions, please let me know.

## Features

This extension controls the RGB lights on your devices to indicate the problems found in your code.

## Requirements

- 64-bit Windows Only
- Requires hardware with ASUS Aura Sync® RGB
- Requires ASUS Aura software to be installed
- Requires LightingService.exe from the ASUS Aura software to be running

## Contributes

### Commands

#### Test connection to Aura (`vscode-aura.testAura`)

This command will test the connection to the Aura Sync software. The command will return a message indicating if the connection was successful and the connected devices will display white for a short time.

#### Close connection to Aura (`vscode-aura.closeAura`)

This command will close the connection to the Aura Sync software. This releases control and allows the lights to return to their default behaviour.

### Configuration

<details>
<summary><h4 style="display:inline">Colo(u)rs (<code>vscode-aura.colors</code>)</h4>

This section contains the configuration for the colo(u)rs used to indicate the problems found in your code.
</summary>

##### Colors > Info (`vscode-aura.colors.info`)

This is the colo(u)r used to indicate an info problem.

*Accepts: hex color string*


##### Colors > Warning (`vscode-aura.colors.warning`)

This is the colo(u)r used to indicate an info problem.

*Accepts: hex color string*

##### Colors > Error (`vscode-aura.colors.error`)

This is the colo(u)r used to indicate an info problem.

*Accepts: hex color string*
</details>


<details>
<summary>
<h4 style="display:inline">Timings (<code>vscode-aura.timings</code>)</h4>

This section contains the configuration for the timings used to indicate the problems found in your code.
</summary>

##### Timings > Update Interval (`vscode-aura.timings.updateInterval`)

This is the length in milliseconds to update the problem count.

*Accepts: number >= 0*


##### Timings > Info Blink (`vscode-aura.timings.infoBlink`)

This is the length in milliseconds for each blink when showing info problems.

*Accepts: number >= 0*


##### Timings > Info Hold (`vscode-aura.timings.infoHold`)

This is the length in milliseconds to hold the light when showing info problems over the high threshold.

*Accepts: number >= 0*


##### Timings > Warning Blink (`vscode-aura.timings.warningBlink`)

This is the length in milliseconds for each blink when showing warning problems.

*Accepts: number >= 0*


##### Timings > Warning Hold (`vscode-aura.timings.warningHold`)

This is the length in milliseconds to hold the light when showing warning problems over the high threshold.

*Accepts: number >= 0*


##### Timings > Error Blink (`vscode-aura.timings.errorBlink`)

This is the length in milliseconds for each blink when showing error problems.

*Accepts: number >= 0*


##### Timings > Error Hold (`vscode-aura.timings.errorHold`)

This is the length in milliseconds to hold the light when showing error problems over the high threshold.

*Accepts: number >= 0*
</details>


<details>
<summary>
<h4 style="display:inline">Thresholds (<code>vscode-aura.thresholds</code>)</h4>

This section contains the configuration for the thresholds when indicating the problems found in your code.
</summary>

##### Thresholds > Info Minimum (`vscode-aura.thresholds.infoMinimum`)

This is the minimum number of info problems needed to start blinking.

*Accepts: number >= 1*


##### Thresholds > Info Maximum (`vscode-aura.thresholds.infoMaximum`)

This is the maximum number of info problems needed before the light will hold for some time. **(If it is below the minimum, holding will always happen!)**

*Accepts: number >= 1*


##### Thresholds > Warning Minimum (`vscode-aura.thresholds.warningMinimum`)

This is the minimum number of warning problems needed to start blinking.

*Accepts: number >= 1*


##### Thresholds > Warning Maximum (`vscode-aura.thresholds.warningMaximum`)

This is the maximum number of warning problems needed before the light will hold for some time. **(If it is below the minimum, holding will always happen!)**

*Accepts: number >= 1*


##### Thresholds > Error Minimum (`vscode-aura.thresholds.errorMinimum`)

This is the minimum number of error problems needed to start blinking.

*Accepts: number >= 1*

##### Thresholds > Error Maximum (`vscode-aura.thresholds.errorMaximum`)

This is the maximum number of error problems needed before the light will hold for some time. **(If it is below the minimum, holding will always happen!)**

*Accepts: number >= 1*
</details>


#### Only Visible Files

This setting determines if the extension will only show problems for the currently visible files.

*Accepts: boolean*


## Release Notes

See the [CHANGELOG](CHANGELOG.md) for more information.


### 0.0.8 - "Pocket Dimension"

Update to Electron abi106.

### 0.0.7 - "Read all about it"

Documents commands and configuration in the README.

### 0.0.6 - "Don't be negative!"

No negative numbers in number input fields.

### 0.0.5 - "Configs: Part Deux"

Option to only show problems in visible files.

### 0.0.4 - "Configs: Prelude"

Light behaviour is now configurable.

### 0.0.3 - "Monopoly: Microsoft Edition"

Edited publish rules and manifest to only allow win32-x64

### 0.0.2 - "Time based check"

Uses time based check rather than using diagnostic event.

### 0.0.1 - "Initial Release"

Initial release.


CHANGELOG.md
# Changelog

All notable changes to the "vscode-aura" extension will be documented in this file.

## [Unreleased]

### Planned

- Config: option - set default light behaviour when no problems

## [0.0.8] - 20/09/2022

- Update to Electron abi106

## [0.0.7] - 10/08/2022

### Added

- README: document commands
- README: document configuration

## [0.0.6] - 10/08/2022

### Fixed

- A bug allowing users to enter negative numbers in the number input fields

## [0.0.5] - 09/08/2022

### Added

- Config: toggle - only show problems for current file

## [0.0.4] - 08/08/2022

### Added

- Config: inputs - set colours for each problem type
- Config: inputs - set timings for each problem type
- Config: inputs - set minimum thresholds for each problem type and the maximum amount before turning from blinking to solid

## [0.0.3] - 30/07/2022

### Added

- Edited publish rules and manifest to only allow win32-x64

## [0.0.2] - 30/07/2022

### Added

- Uses time based check rather than using diagnostic event, which would cause overlap

## [0.0.1] - 30/07/2022

### Added

- Setup dependencies to work in dev and prod
- Setup test and disconnect commands
- Blink the lights based on error, warning, and info count
- Setup the extension's branding and README

[Unreleased]: https://github.com/obfuscatedgenerated/vscode-aura/compare/v0.0.7...HEAD
[0.0.1]: https://github.com/obfuscatedgenerated/vscode-aura/releases/tag/v0.0.1
[0.0.2]: https://github.com/obfuscatedgenerated/vscode-aura/releases/tag/v0.0.2
[0.0.3]: https://github.com/obfuscatedgenerated/vscode-aura/releases/tag/v0.0.3
[0.0.4]: https://github.com/obfuscatedgenerated/vscode-aura/releases/tag/v0.0.4
[0.0.5]: https://github.com/obfuscatedgenerated/vscode-aura/releases/tag/v0.0.5
[0.0.6]: https://github.com/obfuscatedgenerated/vscode-aura/releases/tag/v0.0.6
[0.0.7]: https://github.com/obfuscatedgenerated/vscode-aura/releases/tag/v0.0.7

src/test/runTest.ts
import * as path from 'path';

import { runTests } from '@vscode/test-electron';

const winax = require("winax");

async function main() {
	if(process.arch !== "x64" || process.platform !== "win32") {
		console.error("Tests only run on 64-bit Windows.");
		process.exit(1);
	}

	try {
		let test_aura = new winax.Object("aura.sdk.1");
		winax.release(test_aura);
	} catch (e) {
		if (e instanceof Error && e.message.includes("Invalid class string")) {
			console.error("Tests require Aura installed.");
			process.exit(1);
		} else {
			throw e;
		}
	}
	try {
		// The folder containing the Extension Manifest package.json
		// Passed to `--extensionDevelopmentPath`
		const extensionDevelopmentPath = path.resolve(__dirname, '../../');

		// The path to test runner
		// Passed to --extensionTestsPath
		const extensionTestsPath = path.resolve(__dirname, './suite/index');

		// Download VS Code, unzip it and run the integration test
		await runTests({ extensionDevelopmentPath, extensionTestsPath });
	} catch (err) {
		console.error('Failed to run tests');
		process.exit(1);
	}
}

main();


src/test/suite/index.ts
import * as path from 'path';
import * as Mocha from 'mocha';
import * as glob from 'glob';

export function run(): Promise<void> {
	// Create the mocha test
	const mocha = new Mocha({
		ui: 'tdd',
		color: true,
		timeout: 3000
	});

	const testsRoot = path.resolve(__dirname, '..');

	return new Promise((c, e) => {
		glob('**/**.test.js', { cwd: testsRoot }, (err, files) => {
			if (err) {
				return e(err);
			}

			// Add files to the test suite
			files.forEach(f => mocha.addFile(path.resolve(testsRoot, f)));

			try {
				// Run the mocha test
				mocha.run(failures => {
					if (failures > 0) {
						e(new Error(`${failures} tests failed.`));
					} else {
						c();
					}
				});
			} catch (err) {
				console.error(err);
				e(err);
			}
		});
	});
}


src/test/suite/extension.test.ts
import * as assert from "assert";
import * as vscode from "vscode";

suite("Extension Tests", () => {
	vscode.window.showInformationMessage("Begin tests: extension");
	

	test("Test command: testAura", async () => {
		await vscode.commands.executeCommand("vscode-aura.testAura");
	});

	test("Test command: closeAura", async () => {
		await vscode.commands.executeCommand("vscode-aura.closeAura");
	});
});


src/types.ts
// GENERIC TYPES



// AURA SDK TYPES

export interface IAuraSdk {
    SwitchMode(): void,
    Enumerate(device_type: number): IAuraSyncDeviceCollection,
    ReleaseControl(reserved: number): void,
}

export interface IAuraCollection {
    _NewEnum(): unknown,
    Item(index: number): unknown,
    Count: number,
}

export interface IAuraSyncDeviceCollection extends Omit<IAuraCollection, "Item"> {
    Item(index: number): IAuraSyncDevice,
}

export interface IAuraSyncDevice {
    Lights: IAuraRgbLightCollection,
    Type: number,
    Name: string,
    Width: number,
    Height: number,
    Apply(): void,
}

export interface IAuraSyncKeyboard extends IAuraSyncDevice {
    Keys: IAuraRgbKeyCollection,
    Key(): IAuraRgbKey,
}

export interface IAuraRgbLightCollection extends Omit<IAuraCollection, "Item"> {
    Item(index: number): IAuraRgbLight,
}

export interface IAuraRgbKeyCollection extends Omit<IAuraCollection, "Item"> {
    Item(index: number): IAuraRgbKey,
}

export interface IAuraRgbLight {
    Red: number,
    Green: number,
    Blue: number,
    Name: string,
    Color: number,
}

export interface IAuraRgbKey extends IAuraRgbLight {
    Code: number,
    X: number,
    Y: number,
}


// CONFIGURATION TYPES

export interface IColorConfig {
    info: string,
    warning: string,
    error: string,
}

export interface ITimingConfig {
    updateInterval: number,
    infoBlink: number,
    infoHold: number,
    warningBlink: number,
    warningHold: number,
    errorBlink: number,
    errorHold: number,
}

export interface IThresholdConfig {
    infoMinimum: number,
    infoHigh: number,
    warningMinimum: number,
    warningHigh: number,
    errorMinimum: number,
    errorHigh: number,
}

src/aura_sdk.ts
const winax = require("winax");

import { IAuraSdk, IAuraSyncDeviceCollection, IAuraSyncDevice, IAuraSyncKeyboard, IAuraRgbLight, IAuraRgbKey } from "./types";

let aura: IAuraSdk;
let devices: IAuraSyncDeviceCollection;

let ready = false;

/**
 * Error fired when the Aura SDK is not ready.
 * @date 31/07/2022 - 00:10:01
 *
 * @export
 * @class AuraNotInitError
 * @typedef {AuraNotInitError}
 * @extends {Error}
 */
export class AuraNotInitError extends Error {
    /**
     * Creates an instance of AuraNotInitError.
     * @date 31/07/2022 - 00:10:01
     *
     * @constructor
     */
    constructor() {
        super("Aura SDK is not initialized");
    }
};
/**
 * Error fired when the Aura SDK isn't available on the device.
 * @date 31/07/2022 - 00:10:01
 *
 * @export
 * @class AuraNotFoundError
 * @typedef {AuraNotFoundError}
 * @extends {Error}
 */
export class AuraNotFoundError extends Error {
    /**
     * Creates an instance of AuraNotFoundError.
     * @date 31/07/2022 - 00:10:01
     *
     * @constructor
     */
    constructor() {
        super("Aura SDK not found. Please make sure you have Aura installed and LightingService.exe is running.");
    }
};

/**
 * Initializes the Aura SDK.
 * @date 31/07/2022 - 00:10:01
 */
export let init = (): void => {
    if (ready) {
        console.warn("Aura SDK is already initialized");
        return;
    }

    try {
        aura = new winax.Object("aura.sdk.1");
    } catch (e) {
        if (e instanceof Error && e.message.includes("Invalid class string")) {
            throw new AuraNotFoundError();
        } else {
            throw e;
        }
    }

    aura.SwitchMode();
    devices = aura.Enumerate(0);
    ready = true;
};

/**
 * Lists all Aura devices.
 * @date 31/07/2022 - 00:10:01
 *
 * @returns {(IAuraSyncDeviceCollection | void)}
 */
export let get_devices = (): IAuraSyncDeviceCollection | void => {
    if (!ready) {
        throw new AuraNotInitError();
    }

    return devices;
};

/**
 * Sets all Aura devices to the given (hexadecimal) color.
 * @date 31/07/2022 - 00:10:01
 *
 * @param {number} color
 */
export let set_all_to_color = (color: number): void => {
    if (!ready) {
        throw new AuraNotInitError();
    }

    //for (let dev in devices) {
    for (let i = 0; i < devices.Count; i++) {
        let dev: IAuraSyncDevice = devices.Item(i);
        if (dev.Type === 0x80000) {
            let kb = dev as IAuraSyncKeyboard;
            //for (let key in dev.Keys) {
            for (let j = 0; j < kb.Keys.Count; j++) {
                let key: IAuraRgbKey = kb.Keys.Item(j);
                key.Color = color;
            }
        } else {
            //for (let light in dev.Lights) {
            if (dev.Lights === undefined) {
                console.warn("Device " + dev.Name + " has no lights nor keys");
                continue;
            }
            for (let j = 0; j < dev.Lights.Count; j++) {
                let light: IAuraRgbLight = dev.Lights.Item(j);
                light.Color = color;
            }
            dev.Apply();
        }
    }
};

/**
 * Converts RGB values to a hexadecimal color number.
 * @date 31/07/2022 - 00:10:01
 *
 * @param {number} r
 * @param {number} g
 * @param {number} b
 * @returns {number}
 */
export let rgb_to_color = (r: number, g: number, b: number): number => {
    return b << 16 | g << 8 | r;
};

/**
 * Releases control of all Aura devices.
 * @date 31/07/2022 - 00:10:01
 */
export let close = (): void => {
    if (!ready) {
        throw new AuraNotInitError();
    }

    ready = false;
    aura.ReleaseControl(0);
    winax.release(aura);
};

src/extension.ts
import * as vscode from "vscode";

import * as aura from "./aura_sdk";
import { IColorConfig, ITimingConfig, IThresholdConfig } from "./types";

let connected = false;

export function activate(context: vscode.ExtensionContext) {

	try {
		aura.init();
		connected = true;
	} catch (e) {
		console.error(e);
		if (e instanceof Error) {
			vscode.window.showErrorMessage(e.message);
		}
	}

	console.log("\"vscode-aura\" is now active.");

	aura.set_all_to_color(0);


	context.subscriptions.push(
		vscode.commands.registerCommand("vscode-aura.testAura", async () => {
			if (connected) {
				aura.set_all_to_color(aura.rgb_to_color(255, 255, 255));
				await sleep(1000);
				aura.set_all_to_color(0);
				vscode.window.showInformationMessage("Aura is connected!");
			} else {
				vscode.window.showInformationMessage("Aura is not connected!");
			}
		})
	);

	context.subscriptions.push(
		vscode.commands.registerCommand("vscode-aura.closeAura", () => {
			aura.close();
			connected = false;
			vscode.window.showInformationMessage("Closed connection to Aura. Reload the extension to reconnect.");
		})
	);
	intervalUpdate();
}

export function deactivate() {
	if (connected) {
		aura.close();
	}
}

const sleep = (milliseconds: number) => {
	return new Promise(resolve => setTimeout(resolve, milliseconds));
};

async function flash(color: number, times: number, duration: number) {
	if (times > 0) {
		await sleep(duration);
		aura.set_all_to_color(color);

		await sleep(duration);
		aura.set_all_to_color(0);

		await flash(color, times - 1, duration);
	} else {
		await sleep(duration);
		return Promise.resolve();
	}
}

let lastCounts = [0, 0, 0, 0];

async function intervalUpdate() {
	if (connected) {
		let config = vscode.workspace.getConfiguration("vscode-aura");
		let rawTime = config.get("timings.updateInterval", 3000) as any;
		let time = (typeof rawTime === "number") ? rawTime as number : resetNum(config, "timings", "updateInterval", rawTime) as number;

		await updateLight();
		await sleep(time);
		intervalUpdate();
	}
}

function convertForAura(color: string) {
	if (color.startsWith("#")) { // remove leading #
		color = color.substring(1);
	}

	if (color.length === 4) { // if RGBA, convert to RGB
		color = color.substring(0, 5);
	}
	if (color.length === 8) { // if RRGGBBAA, convert to RRGGBB
		color = color.substring(0, 7);
	}
	if (color.length === 3) { // if RGB, convert to RRGGBB
		color = color.split("").map(c => c + c).join("");
	}

	return color.substring(4, 6) + color.substring(2, 4) + color.substring(0, 2); // rearrange to BBGGRR
}

function resetColor(config: any, k: string, v: string) {
	vscode.window.showErrorMessage(`Invalid color ${v} for ${k}. Resetting to default...`);

	let def = config.inspect("colors." + k)?.defaultValue as string;
	let globalValue = config.inspect("colors." + k)?.globalValue;
	let workspaceFolderValue = config.inspect("colors." + k)?.workspaceFolderValue;
	let workspaceValue = config.inspect("colors." + k)?.workspaceValue;

	if (def) {
		if (globalValue === v) {
			config.update("colors." + k, def, vscode.ConfigurationTarget.Global);
		} else if (workspaceFolderValue === v) {
			config.update("colors." + k, def, vscode.ConfigurationTarget.WorkspaceFolder);
		} else if (workspaceValue === v) {
			config.update("colors." + k, def, vscode.ConfigurationTarget.Workspace);
		}
		return parseInt(convertForAura(def), 16);
	} else {
		if (globalValue === v) {
			config.update("colors." + k, "#FFFFFF", vscode.ConfigurationTarget.Global);
		} else if (workspaceFolderValue === v) {
			config.update("colors." + k, "#FFFFFF", vscode.ConfigurationTarget.WorkspaceFolder);
		} else if (workspaceValue === v) {
			config.update("colors." + k, "#FFFFFF", vscode.ConfigurationTarget.Workspace);
		}
		return parseInt("FFFFFF", 16);
	}
}

function parseColors(colors: IColorConfig, config: any) {
	return Object.fromEntries(
		Object.entries(colors).map(([k, v]) => {
			let val: number;
			if (typeof v === "string") {
				val = parseInt(convertForAura(v), 16);
				if (isNaN(val)) {
					val = resetColor(config, k, v);
				}
			} else {
				val = resetColor(config, k, v);
			}
			return [k as keyof typeof colors, val];
		})
	);
}

function resetNum(config: any, type: string, k: string, v: any) {
	vscode.window.showErrorMessage(`Invalid number ${v} for ${k}. Resetting to default...`);

	let def = config.inspect(type + "." + k)?.defaultValue as number;
	let globalValue = config.inspect(type + "." + k)?.globalValue;
	let workspaceFolderValue = config.inspect(type + "." + k)?.workspaceFolderValue;
	let workspaceValue = config.inspect(type + "." + k)?.workspaceValue;

	if (def) {
		if (globalValue === v) {
			config.update(type + "." + k, def, vscode.ConfigurationTarget.Global);
		} else if (workspaceFolderValue === v) {
			config.update(type + "." + k, def, vscode.ConfigurationTarget.WorkspaceFolder);
		} else if (workspaceValue === v) {
			config.update(type + "." + k, def, vscode.ConfigurationTarget.Workspace);
		}
		return def;
	} else {
		if (globalValue === v) {
			config.update(type + "." + k, 200, vscode.ConfigurationTarget.Global);
		} else if (workspaceFolderValue === v) {
			config.update(type + "." + k, 200, vscode.ConfigurationTarget.WorkspaceFolder);
		} else if (workspaceValue === v) {
			config.update(type + "." + k, 200, vscode.ConfigurationTarget.Workspace);
		}
		return 200;
	}
}

function validateNums(config: any, type: string, values: ITimingConfig | IThresholdConfig) {
	return Object.fromEntries(
		Object.entries(values).map(([k, v]) => {
			let val: number;
			if (typeof v === "number" && v >= 0) {
				val = v;
			} else {
				val = resetNum(config, type, k, v);
			}
			return [k as keyof typeof values, val];
		})
	);
}

function resetBool(config: any, k: string, v: any) {
	vscode.window.showErrorMessage(`Invalid boolean ${v} for ${k}. Resetting to default...`);

	let def = config.inspect(k)?.defaultValue as boolean;
	let globalValue = config.inspect(k)?.globalValue;
	let workspaceFolderValue = config.inspect(k)?.workspaceFolderValue;
	let workspaceValue = config.inspect(k)?.workspaceValue;

	if (def) {
		if (globalValue === v) {
			config.update(k, def, vscode.ConfigurationTarget.Global);
		} else if (workspaceFolderValue === v) {
			config.update(k, def, vscode.ConfigurationTarget.WorkspaceFolder);
		} else if (workspaceValue === v) {
			config.update(k, def, vscode.ConfigurationTarget.Workspace);
		}
		return def;
	} else {
		if (globalValue === v) {
			config.update(k, false, vscode.ConfigurationTarget.Global);
		} else if (workspaceFolderValue === v) {
			config.update(k, false, vscode.ConfigurationTarget.WorkspaceFolder);
		} else if (workspaceValue === v) {
			config.update(k, false, vscode.ConfigurationTarget.Workspace);
		}
		return false;
	}
}

function validateBool(config: any, k: string, v: any) {
	let val: boolean;
	if (typeof v === "boolean") {
		val = v;
	} else {
		val = resetBool(config, k, v);
	}
	return val;
}

async function updateLight() {
	let config = vscode.workspace.getConfiguration("vscode-aura");
	let diag = vscode.languages.getDiagnostics();

	if (validateBool(config, "onlyVisibleFiles", config.get("onlyVisibleFiles", false))) {
		diag = diag.filter(([uri, d]) => {
			if (vscode.window.visibleTextEditors.find(e => e.document.uri === uri) || vscode.window.visibleNotebookEditors.find(e => e.notebook.uri === uri)) {
				return true;
			} else {
				return false;
			}
		});
	}

	let colors = config.get("colors") as IColorConfig;
	let parsedColors = parseColors(colors, config);
	let timings = validateNums(config, "timings", config.get("timings") as ITimingConfig);
	let thresholds = validateNums(config, "thresholds", config.get("thresholds") as IThresholdConfig);

	let errors = 0;
	let warnings = 0;
	let infos = 0;
	let other = 0;

	for (let i in diag) {
		let notices = diag[i][1];
		for (let j in notices) {
			let notice = notices[j];
			if (notice.severity === vscode.DiagnosticSeverity.Error) {
				errors++;
			} else if (notice.severity === vscode.DiagnosticSeverity.Warning) {
				warnings++;
			} else if (notice.severity === vscode.DiagnosticSeverity.Information) {
				infos++;
			} else {
				other++;
			}
		}
	}

	if ([errors, warnings, infos, other].join() === lastCounts.join()) {
		return;
	}

	lastCounts = [errors, warnings, infos, other];

	if (errors >= thresholds.errorMinimum) {
		if (errors < thresholds.errorHigh) {
			await flash(parsedColors.error, errors, timings.errorBlink);
		} else {
			aura.set_all_to_color(parsedColors.error);
			await sleep(timings.errorHold);
			aura.set_all_to_color(0);
		}
	}

	if (warnings >= thresholds.warningMinimum) {
		if (warnings < thresholds.warningHigh) {
			await flash(parsedColors.warning, warnings, timings.warningBlink);
		} else {
			aura.set_all_to_color(parsedColors.warning);
			await sleep(timings.warningHold);
			aura.set_all_to_color(0);
		}
	}

	if (infos >= thresholds.infoMinimum) {
		if (infos < thresholds.infoHigh) {
			await flash(parsedColors.info, infos, timings.infoBlink);
		} else {
			aura.set_all_to_color(parsedColors.info);
			await sleep(timings.infoHold);
			aura.set_all_to_color(0);
		}
	}
}